crates/driver/src/domain/competition/score.rs

```diff
@@ -76,31 +89,102 @@ impl From<eth::U256> for ObjectiveValue {
-    ObjectiveValueNonPositive,
+    /// The solution has zero score. Zero score solutions are not allowed as per
+    /// CIP20 definition. The main reason being that reference score is zero,
+    /// and if only one solution is in competition with zero score, that
+    /// solution would receive 0 reward (reward = score - reference score).
     #[error("score is zero")]
     ZeroScore,
-    #[error("objective value {0:?} is higher than the objective {1:?}")]
-    ScoreHigherThanObjective(Score, ObjectiveValue),
-    #[error("success probability is out of range {0:?}")]
-    SuccessProbabilityOutOfRange(SuccessProbability),
-    #[error("invalid objective value")]
+    /// Protocol does not allow solutions that are claimed to be "better" than
+    /// the actual value they bring (quality). It is expected that score
+    /// is always lower than quality, because there is always some
+    /// execution cost that needs to be incorporated into the score and lower
+    /// it.
+    #[error("score {0:?} is higher than the quality {1:?}")]
+    ScoreHigherThanQuality(Score, Quality),
+    /// Errors only applicable to scores that use success probability.
+    #[error(transparent)]
+    RiskAdjusted(#[from] risk::Error),
+    #[error(transparent)]
     Boundary(#[from] boundary::Error),
 }
+
+pub mod risk {
+    /// Contains functionality and error types for scores that are based on
+    /// success probability.
+    use {
+        super::{ObjectiveValue, Score},
+        crate::{boundary, domain::eth},
+        bigdecimal::Zero,
+    };
+
+    /// Constructs a score based on the success probability of a solution.
+    impl Score {
+        pub fn new(
+            score_cap: Score,
+            objective_value: ObjectiveValue,
+            success_probability: SuccessProbability,
+            failure_cost: eth::GasCost,
+        ) -> Result<Self, Error> {
+            if objective_value.is_zero() {
+                return Err(Error::ObjectiveValueNonPositive);
+            }
+            if !(0.0..=1.0).contains(&success_probability.0) {
+                return Err(Error::SuccessProbabilityOutOfRange(success_probability));
+            }
+            Ok(boundary::score::score(
+                score_cap,
+                objective_value,
+                success_probability,
+                failure_cost,
+            )?)
+        }
+    }
+
+    /// Represents the probability that a solution will be successfully settled.
+    #[derive(Debug, Copy, Clone)]
+    pub struct SuccessProbability(pub f64);
+
+    impl From<f64> for SuccessProbability {
+        fn from(value: f64) -> Self {
+            Self(value)
+        }
+    }
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Error {
+        /// Solution has success probability that is outside of the allowed
+        /// range [0, 1]
+        #[error("success probability is out of range {0:?}")]
+        SuccessProbabilityOutOfRange(SuccessProbability),
```

Comment:
The existence of this error is very weird to me, as it shouldn't be called a probability in the first place if it isn't in this range. How is it possible that we get a nonsense probability?

Reply:
This could be handled by implementing `TryFrom` instead of `From`. Not sure if this would make the error handling more annoying because the transformation from `f64` to `SuccessProbability` might happen where we don't have the required types to report this error back.

Reply:
https://github.com/cowprotocol/services/pull/2022/commits/39858e03f8c9b0bcdec8ead269ae5f094584db78